### PR TITLE
Add public APIs to easily get ControlFlowGraph for local functions/an…

### LINKED ISSUE #29836: Semantics of GetAnonymousFunctionControlFlowGraph and GetLocalFunctionControlFlowGraph
[ControlFlowGraph.GetAnonymousFunctionControlFlowGraph](http://source.roslyn.io/#q=GetAnonymousFunctionControlFlowGraph) and [ControlFlowGraph.GetLocalFunctionControlFlowGraph](http://source.roslyn.io/#q=GetLocalFunctionControlFlowGraph) currently work only if the input (`IFlowAnonymousFunctionOperation` or the local function `IMethodSymbol`) is defined in the method for the corresponding flow graph. However, if the local function or lambda is defined in an even outer scope, it does not work.

For example, consider the below source:
```
class Test
{
    void M1()
    {
        MyLocalFunction1();
        return;

        void MyLocalFunction1()
        {
            MyLocalFunction2();  // Need the control flow graph for invoked local function here.
        };

        void MyLocalFunction2()
        {
        };
    }
}
```

While analyzing the CFG for `MyLocalFunction1`, when we run into invocation `MyLocalFunction2()`, attempting to invoke `GetLocalFunctionControlFlowGraph` on MyLocalFunction1's CFG fails. This forces the analysis to always track the chain of parent control flow graphs to get the correct local function flow graph:
```
       public ControlFlowGraph GetLocalFunctionControlFlowGraph(IMethodSymbol localFunction)
             => _controlFlowGraph.LocalFunctions.Contains(localFunction) ?
                _controlFlowGraph.GetLocalFunctionControlFlowGraph(localFunction):
                _parentControlFlowGraphOpt?.GetLocalFunctionControlFlowGraph(localFunction);
```

The implementation is even more messier for getting the flow graph for anonymous functions/lambda, as we do not expose the defined lambdas on `ControlFlowGraph`:
```
        public ControlFlowGraph GetAnonymousFunctionControlFlowGraph(IFlowAnonymousFunctionOperation lambda)
        {
            try
            {
                return _controlFlowGraph.GetAnonymousFunctionControlFlowGraph(lambda);
            }
            catch (ArgumentOutOfRangeException)
            {
                return _parentControlFlowGraphOpt?.GetAnonymousFunctionControlFlowGraph(lambda);
            }
        }
```

**Feature Requests:**
1. Either change the semantics of `GetAnonymousFunctionControlFlowGraph` and `GetLocalFunctionControlFlowGraph` to walk up the parent chain of control flow graphs OR expose new helper methods on `ControlFlowGraph` to do the walk. I personally prefer the former, i.e. change semantics of existing APIs, unless we see big performance concerns.
2. Expose `ControlFlowGraph.FlowAnonymousFunctionOperations` similar to `ControlFlowGraph.LocalFunctions` which we currently expose.

### PR DESCRIPTION
**Pull request by mavasani**
…onymous functions in scope.

1. Exposes `ControlFlowGraph.Parent`
2. Expose extension methods `ControlFlowGraph.GetLocalFunctionControlFlowGraphInScope` and `ControlFlowGraph.GetAnonymousFunctionControlFlowGraphInScope` to walk up the CFG parent chain to find the local function/anonymous function flow graphs in scope.

Fixes #29836

### CODE REVIEW

File: src/Compilers/Core/Portable/Operations/ControlFlowGraph.cs [hunk 1]
```diff
@@ -193,6 +196,13 @@ internal static ControlFlowGraph CreateCore(IOperation operation, string argumen
         /// </summary>
         public IOperation OriginalOperation { get; }
 
+        /// <summary>
+        /// Optional parent control flow graph for this graph.
+        /// Non-null for a control flow graph generated for a local function or a lambda.
+        /// Null, otherwise.
```
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraph.cs]: Nit: drop the comma.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraph.cs [hunk 2]
```diff
@@ -220,11 +230,22 @@ public ControlFlowGraph GetLocalFunctionControlFlowGraph(IMethodSymbol localFunc
...
-            if (!_localFunctionsMap.TryGetValue(localFunction, out (ControlFlowRegion enclosing, ILocalFunctionOperation operation, int ordinal) info))
+            if (!TryGetLocalFunctionControlFlowGraph(localFunction, cancellationToken, out var controlFlowGraph))
             {
                 throw new ArgumentOutOfRangeException(nameof(localFunction));
             }
 
+            return controlFlowGraph;
+        }
+
+        internal bool TryGetLocalFunctionControlFlowGraph(IMethodSymbol localFunction, CancellationToken cancellationToken, out ControlFlowGraph controlFlowGraph)
```
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraph.cs]: Am I missing something, or is `TryGetAnonymousFunctionControlFlowGraph` basically the same function, except over a different map? Consider refactoring with a helper that takes a ref to the correct graph.

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IAnonymousFunctionExpression.cs [hunk 3]
```diff
@@ -753,24 +753,91 @@ void M(System.Action d1, System.Action<bool, bool> d2)
...
             IFlowAnonymousFunctionOperation lambdaD1 = getLambda(graphM);
 
             Assert.Throws<ArgumentNullException>(() => graphM.GetLocalFunctionControlFlowGraph(null));
             Assert.Throws<ArgumentOutOfRangeException>(() => graphM.GetLocalFunctionControlFlowGraph(lambdaD1.Symbol));
+            Assert.Throws<ArgumentNullException>(() => graphM.GetLocalFunctionControlFlowGraphInScope(null));
+            Assert.Throws<ArgumentOutOfRangeException>(() => graphM.GetLocalFunctionControlFlowGraphInScope(lambdaD1.Symbol));
 
             var graphD1 = graphM.GetAnonymousFunctionControlFlowGraph(lambdaD1);
             Assert.NotNull(graphD1);
+            Assert.Same(graphM, graphD1.Parent);
```
Comment (333fred) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IAnonymousFunctionExpression.cs]: Consider adding a similar set of assertions for local functions.

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IAnonymousFunctionExpression.cs [hunk 4]
```diff
@@ -753,24 +753,91 @@ void M(System.Action d1, System.Action<bool, bool> d2)
...
+            var graphD1_FromExtension = graphM.GetAnonymousFunctionControlFlowGraphInScope(lambdaD1);
+            Assert.Same(graphD1, graphD1_FromExtension);
 
             IFlowAnonymousFunctionOperation lambdaD2 = getLambda(graphD1);
+            var graphD2 = graphD1.GetAnonymousFunctionControlFlowGraph(lambdaD2);
```
Comment (333fred) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IAnonymousFunctionExpression.cs]: Same here.

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IAnonymousFunctionExpression.cs [hunk 5]
```diff
@@ -753,24 +753,91 @@ void M(System.Action d1, System.Action<bool, bool> d2)
...
 
             IFlowAnonymousFunctionOperation getLambda(ControlFlowGraph graph)
             {
                 return graph.Blocks.SelectMany(b => b.Operations.SelectMany(o => o.DescendantsAndSelf())).OfType<IFlowAnonymousFunctionOperation>().Single();
             }
         }
+
+        [CompilerTrait(CompilerFeature.IOperation, CompilerFeature.Dataflow)]
+        [Fact]
+        public void LambdaFlow_05()
```
Comment (333fred) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IAnonymousFunctionExpression.cs]: Is it interesting to replicate this test with local functions?

Comment (333fred) [reply src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IAnonymousFunctionExpression.cs]: Ah, nevermind, I thought I was seeing all of the test files. Ignore.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraph.cs (hunk 2 above)
Comment (mavasani) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraph.cs]: I am not sure I understand your suggestion. `TryGetLocalFunctionControlFlowGraph` is basically the just refactoring out the existing code from `GetLocalFunctionControlFlowGraph`, but does not throw any exception if it does not succeed. This function is used in `GetLocalFunctionControlFlowGraph` and newly added `GetLocalFunctionControlFlowGraphInScope`.

---
In reply to: [223788574](https://github.com/dotnet/roslyn/pull/30395#discussion_r223788574) [](ancestors = 223788574)

[the remark this answers is not in the document]
Comment (mavasani) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraph.cs]: Tracked by [https://github.com/dotnet/roslyn/issues/28130](https://github.com/dotnet/roslyn/issues/28130)

---
In reply to: [223792375](https://github.com/dotnet/roslyn/pull/30395#discussion_r223792375) [](ancestors = 223792375)

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IAnonymousFunctionExpression.cs (hunk 3 above)
Comment (mavasani) [reply src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IAnonymousFunctionExpression.cs]: `LocalFunctionFlow_11`

---
In reply to: [223793329](https://github.com/dotnet/roslyn/pull/30395#discussion_r223793329) [](ancestors = 223793329)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraph.cs (hunk 2 above)
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraph.cs]: `TryGetLocalFunctionControlFlowGraph` and `TryGetAnonymousFunctionControlFlowGraph` are pretty much clones of each other, except that they take different maps. The suggestion is to make one helper that does the actual work and takes a ref to the actual graph (with appropriate generics).

---
In reply to: [223794770](https://github.com/dotnet/roslyn/pull/30395#discussion_r223794770) [](ancestors = 223794770,223788574)

Comment (mavasani) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraph.cs]: I think the implementations are also slightly different. There is additional assert `Debug.Assert(localFunction == LocalFunctions[info.ordinal]);` and access to `LocalFunctions` here versus no assert and access to only `_anonymousFunctionsMap` below. I agree that we can still adjust these to use a single generic method, but I'd let @AlekseyTs know his preference as he initially wrote this code.

---
In reply to: [223804899](https://github.com/dotnet/roslyn/pull/30395#discussion_r223804899) [](ancestors = 223804899,223794770,223788574)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+namespace Microsoft.CodeAnalysis.FlowAnalysis
+{
+    public static partial class ControlFlowGraphExtensions
+    {
+        /// <summary>
+        /// Gets or creates a control flow graph for the given <paramref name="localFunction"/> defined in
+        /// the given <paramref name="controlFlowGraph"/> or any of it's parent control flow graphs.
+        /// </summary>
+        public static ControlFlowGraph GetLocalFunctionControlFlowGraphInScope(this ControlFlowGraph controlFlowGraph, IMethodSymbol localFunction, CancellationToken cancellationToken = default)
```
[the remark this answers is not in the document]
Comment (mavasani) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphExtensions.cs]: Tracked by [https://github.com/dotnet/roslyn/issues/28130](https://github.com/dotnet/roslyn/issues/28130)

---
In reply to: [223792484](https://github.com/dotnet/roslyn/pull/30395#discussion_r223792484) [](ancestors = 223792484)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraph.cs (hunk 2 above)
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraph.cs]: Ok. Just a suggestion, not a blocker.

---
In reply to: [223811033](https://github.com/dotnet/roslyn/pull/30395#discussion_r223811033) [](ancestors = 223811033,223804899,223794770,223788574)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+                throw new ArgumentNullException(nameof(localFunction));
+            }
+
+            do
+            {
+                if (controlFlowGraph.TryGetLocalFunctionControlFlowGraph(localFunction, cancellationToken, out ControlFlowGraph localFunctionControlFlowGraph))
+                {
+                    return localFunctionControlFlowGraph;
+                }
+            } while ((controlFlowGraph = controlFlowGraph.Parent) != null);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphExtensions.cs]: >while ((controlFlowGraph = controlFlowGraph.Parent) != null); [](start = 14, length = 61)

I think our coding style requires ```while``` to be on a separate line

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+                throw new ArgumentNullException(nameof(anonymousFunction));
+            }
+
+            do
+            {
+                if (controlFlowGraph.TryGetAnonymousFunctionControlFlowGraph(anonymousFunction, cancellationToken, out ControlFlowGraph localFunctionControlFlowGraph))
+                {
+                    return localFunctionControlFlowGraph;
+                }
+            } while ((controlFlowGraph = controlFlowGraph.Parent) != null);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphExtensions.cs]: >while ((controlFlowGraph = controlFlowGraph.Parent) != null); [](start = 14, length = 61)

Same here

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 1)

**Comment by mavasani** [comment]
@333fred Any further feedback?

**Review by 333fred — APPROVED** [review]
LGTM (commit 2)

**Comment by 333fred** [comment]
@dotnet-bot test roslyn-ci please

**Comment by mavasani** [comment]
retest this please

**Pull request merged**